core/src/main/java/com/linecorp/armeria/internal/common/stream/AppendingPublisher.java

```diff
@@ -0,0 +1,198 @@
+                complete();
+                return;
+            }
+            if (appendNeedToSend) {
+                append();
+                if (n > 1) {
+                    complete();
+                }
+                return;
+            }
+            for (;;) {
+                final long oldRequested = requested;
+                final long newRequested = LongMath.saturatedAdd(oldRequested, n);
+                if (requestedUpdater.compareAndSet(this, oldRequested, newRequested)) {
+                    if (oldRequested > 0) {
+                        return;
+                    }
+                    break;
+                }
+            }
+            if (!subscribed) {
+                subscribed = true;
+                publisher.subscribe(this);
+            }
+            requestUpstream();
+        }
+
+        private void append() {
+            appendSent = true;
+            downstream.onNext(append);
+        }
+
+        private void complete() {
+            completeSent = true;
+            downstream.onComplete();
+        }
+
+        private void requestUpstream() {
+            for (;;) {
+                final Subscription upstream = this.upstream;
+                final long requested = this.requested;
+                if (upstream == null || requested == 0) {
+                    return;
+                }
+                if (requestedUpdater.compareAndSet(this, requested, 0)) {
+                    needToPublishUpdater.addAndGet(this, requested);
+                    upstream.request(requested);
+                }
+            }
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (cancelled) {
+                subscription.cancel();
+                return;
+            }
+            upstream = subscription;
+            requestUpstream();
+        }
+
+        @Override
+        public void cancel() {
+            if (cancelled) {
+                return;
+            }
+            cancelled = true;
+            downstream = NoopSubscriber.get();
+            final Subscription upstream = this.upstream;
+            if (upstream != null) {
+                upstream.cancel();
+            }
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+            needToPublishUpdater.decrementAndGet(this);
+            downstream.onNext(item);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+            if (cancelled) {
+                return;
+            }
+            cancelled = true;
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            final long needToPublish = needToPublishUpdater.get(this);
+            if (needToPublish == 0) {
+                appendNeedToSend = true;
```

Comment:
`append` may not be passed to the `downstream` if the thread that calls `request(n)` passes line 95 at this point.
How about setting a special value (e.g. `-1`) that indicates the flag `needToPublish` and passing the `append` in `requestUpstream()` method?